File: backend/app/services/tag_service.py
"""Service for managing tags"""
from typing import List, Dict, Any, Optional
from uuid import UUID
from ..services.supabase_service import get_supabase_service
from ..utils.logging_config import get_logger

logger = get_logger(__name__)


class TagService:
    """Service for tag operations"""
    
    def __init__(self):
        self.supabase = get_supabase_service()
    
    async def get_tags_by_user(self, user_id: UUID) -> List[Dict[str, Any]]:
        """
        Get all tags for a user with document counts
        
        Args:
            user_id: User UUID
            
        Returns:
            List of tag records with document counts
        """
        try:
            # Get tags
            response = self.supabase.client.table("tags").select("*").eq("user_id", str(user_id)).order("name").execute()
            
            tags = response.data
            
            # Get document counts for each tag
            for tag in tags:
                count_response = self.supabase.client.table("file_tags").select("id", count="exact").eq("tag_id", tag["id"]).execute()
                tag["document_count"] = count_response.count if count_response.count else 0
            
            logger.info(f"Retrieved {len(tags)} tags for user {user_id}")
            return tags
            
        except Exception as e:
            logger.error(f"Error getting tags for user {user_id}: {e}")
            raise
    
    async def create_tag(
        self,
        user_id: UUID,
        name: str,
        color: str = "#2196F3"
    ) -> Dict[str, Any]:
        """
        Create a new tag
        
        Args:
            user_id: User UUID
            name: Tag name
            color: Hex color code
            
        Returns:
            Created tag record
        """
        try:
            # Check for duplicate tag name for this user
            existing = self.supabase.client.table("tags").select("*").eq("user_id", str(user_id)).eq("name", name).execute()
            
            if existing.data:
                raise Exception(f"Tag with name '{name}' already exists")
            
            tag_data = {
                "user_id": str(user_id),
                "name": name,
                "color": color
            }
            
            response = self.supabase.client.table("tags").insert(tag_data).execute()
            
            if response.data:
                tag = response.data[0]
                tag["document_count"] = 0
                logger.info(f"Created tag {tag['id']} for user {user_id}")
                return tag
            else:
                raise Exception("Failed to create tag")
                
        except Exception as e:
            logger.error(f"Error creating tag: {e}")
            raise
    
    async def update_tag(
        self,
        tag_id: UUID,
        user_id: UUID,
        updates: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Update an existing tag
        
        Args:
            tag_id: Tag UUID
            user_id: User UUID
            updates: Fields to update
            
        Returns:
            Updated tag record
        """
        try:
            # Verify ownership
            existing = self.supabase.client.table("tags").select("*").eq("id", str(tag_id)).eq("user_id", str(user_id)).execute()
            
            if not existing.data:
                raise Exception("Tag not found or access denied")
            
            # Check for duplicate name if name is being updated
            if "name" in updates:
                duplicate = self.supabase.client.table("tags").select("*").eq("user_id", str(user_id)).eq("name", updates["name"]).neq("id", str(tag_id)).execute()
                
                if duplicate.data:
                    raise Exception(f"Tag with name '{updates['name']}' already exists")
            
            # Update tag
            response = self.supabase.client.table("tags").update(updates).eq("id", str(tag_id)).execute()
            
            if response.data:
                tag = response.data[0]
                # Get document count
                count_response = self.supabase.client.table("file_tags").select("id", count="exact").eq("tag_id", str(tag_id)).execute()
                tag["document_count"] = count_response.count if count_response.count else 0
                
                logger.info(f"Updated tag {tag_id}")
                return tag
            else:
                raise Exception("Failed to update tag")
                
        except Exception as e:
            logger.error(f"Error updating tag {tag_id}: {e}")
            raise
    
    async def delete_tag(
        self,
        tag_id: UUID,
        user_id: UUID
    ) -> bool:
        """
        Delete a tag and all its file associations
        
        Args:
            tag_id: Tag UUID
            user_id: User UUID
            
        Returns:
            True if deleted successfully
        """
        try:
            # Verify ownership
            existing = self.supabase.client.table("tags").select("*").eq("id", str(tag_id)).eq("user_id", str(user_id)).execute()
            
            if not existing.data:
                raise Exception("Tag not found or access denied")
            
            # Delete file associations first (cascade should handle this, but being explicit)
            self.supabase.client.table("file_tags").delete().eq("tag_id", str(tag_id)).execute()
            
            # Delete tag
            self.supabase.client.table("tags").delete().eq("id", str(tag_id)).execute()
            
            logger.info(f"Deleted tag {tag_id}")
            return True
            
        except Exception as e:
            logger.error(f"Error deleting tag {tag_id}: {e}")
            raise
    
    async def add_tag_to_file(
        self,
        user_id: UUID,
        file_id: str,
        tag_id: UUID
    ) -> Dict[str, Any]:
        """
        Add a tag to a file
        
        Args:
            user_id: User UUID
            file_id: Google Drive file ID
            tag_id: Tag UUID
            
        Returns:
            Created file-tag relationship
        """
        try:
            # Verify tag ownership
            tag = self.supabase.client.table("tags").select("*").eq("id", str(tag_id)).eq("user_id", str(user_id)).execute()
            
            if not tag.data:
                raise Exception("Tag not found or access denied")
            
            # Check if association already exists
            existing = self.supabase.client.table("file_tags").select("*").eq("file_id", file_id).eq("tag_id", str(tag_id)).execute()
            
            if existing.data:
                logger.info(f"Tag {tag_id} already associated with file {file_id}")
                return existing.data[0]
            
            # Create association
            file_tag_data = {
                "user_id": str(user_id),
                "file_id": file_id,
                "tag_id": str(tag_id)
            }
            
            response = self.supabase.client.table("file_tags").insert(file_tag_data).execute()
            
            if response.data:
                logger.info(f"Added tag {tag_id} to file {file_id}")
                return response.data[0]
            else:
                raise Exception("Failed to add tag to file")
                
        except Exception as e:
            logger.error(f"Error adding tag to file: {e}")
            raise
    
    async def remove_tag_from_file(
        self,
        user_id: UUID,
        file_id: str,
        tag_id: UUID
    ) -> bool:
        """
        Remove a tag from a file
        
        Args:
            user_id: User UUID
            file_id: Google Drive file ID
            tag_id: Tag UUID
            
        Returns:
            True if removed successfully
        """
        try:
            # Delete association
            self.supabase.client.table("file_tags").delete().eq("file_id", file_id).eq("tag_id", str(tag_id)).eq("user_id", str(user_id)).execute()
            
            logger.info(f"Removed tag {tag_id} from file {file_id}")
            return True
            
        except Exception as e:
            logger.error(f"Error removing tag from file: {e}")
            raise
    
    async def get_tags_for_file(
        self,
        user_id: UUID,
        file_id: str
    ) -> List[Dict[str, Any]]:
        """
        Get all tags for a file
        
        Args:
            user_id: User UUID
            file_id: Google Drive file ID
            
        Returns:
            List of tags
        """
        try:
            # Get file-tag associations
            file_tags = self.supabase.client.table("file_tags").select("tag_id").eq("file_id", file_id).eq("user_id", str(user_id)).execute()
            
            if not file_tags.data:
                return []
            
            tag_ids = [ft["tag_id"] for ft in file_tags.data]
            
            # Get tag details
            tags = self.supabase.client.table("tags").select("*").in_("id", tag_ids).execute()
            
            logger.info(f"Retrieved {len(tags.data)} tags for file {file_id}")
            return tags.data
            
        except Exception as e:
            logger.error(f"Error getting tags for file {file_id}: {e}")
            raise
    
    async def bulk_tag_files(
        self,
        user_id: UUID,
        file_ids: List[str],
        tag_ids: List[UUID]
    ) -> Dict[str, Any]:
        """
        Apply multiple tags to multiple files
        
        Args:
            user_id: User UUID
            file_ids: List of Google Drive file IDs
            tag_ids: List of tag UUIDs
            
        Returns:
            Result with counts
        """
        tagged_count = 0
        failed_count = 0
        
        try:
            # Verify all tags belong to user
            tags = self.supabase.client.table("tags").select("id").eq("user_id", str(user_id)).in_("id", [str(tid) for tid in tag_ids]).execute()
            
            if len(tags.data) != len(tag_ids):
                raise Exception("One or more tags not found or access denied")
            
            # Create associations
            for file_id in file_ids:
                for tag_id in tag_ids:
                    try:
                        await self.add_tag_to_file(user_id, file_id, tag_id)
                        tagged_count += 1
                    except Exception as e:
                        logger.error(f"Error tagging file {file_id} with tag {tag_id}: {e}")
                        failed_count += 1
            
            result = {
                "success": True,
                "tagged_count": tagged_count,
                "failed_count": failed_count,
                "message": f"Tagged {tagged_count} file-tag pairs, {failed_count} failed"
            }
            
            logger.info(f"Bulk tag completed: {result['message']}")
            return result
            
        except Exception as e:
            logger.error(f"Error during bulk tagging: {e}")
            raise


# Singleton instance
_tag_service = None


def get_tag_service() -> TagService:
    """Get or create tag service instance"""
    global _tag_service
    if _tag_service is None:
        _tag_service = TagService()
    return _tag_service

File: frontend/lib/services/tag_service.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:uuid/uuid.dart';
import 'package:drift/drift.dart';
import '../models/tag.dart' as models;
import '../database/database.dart';
import 'api_service.dart';
import 'auth_service.dart';
import 'connectivity_service.dart';

/// Service for managing tags
class TagService extends ChangeNotifier {
  final AppDatabase _database;
  final ApiService _apiService;
  final AuthService _authService;
  final ConnectivityService _connectivityService;

  // Singleton instance
  static TagService? _instance;
  factory TagService({
    required AppDatabase database,
    required ApiService apiService,
    required AuthService authService,
    required ConnectivityService connectivityService,
  }) {
    _instance ??= TagService._internal(
      database: database,
      apiService: apiService,
      authService: authService,
      connectivityService: connectivityService,
    );
    return _instance!;
  }

  TagService._internal({
    required AppDatabase database,
    required ApiService apiService,
    required AuthService authService,
    required ConnectivityService connectivityService,
  }) : _database = database,
       _apiService = apiService,
       _authService = authService,
       _connectivityService = connectivityService;

  final _uuid = const Uuid();

  /// Get all tags for the current user
  Future<List<models.Tag>> getTags() async {
    try {
      // Try to sync from backend if online
      if (_connectivityService.isOnline) {
        await _syncTagsFromBackend();
      }

      // Get from local cache
      final tagRecords = await _database.select(_database.tags).get();

      // Get document counts
      final tags = <models.Tag>[];
      for (final record in tagRecords) {
        final count = await _getDocumentCount(record.id);
        tags.add(
          models.Tag(
            id: record.id,
            userId: record.userId,
            name: record.name,
            color: record.color,
            createdAt: record.createdAt,
            updatedAt: record.updatedAt,
            documentCount: count,
          ),
        );
      }

      return tags;
    } catch (e) {
      debugPrint('Error getting tags: $e');
      rethrow;
    }
  }

  /// Create a new tag
  Future<models.Tag> createTag({
    required String userId,
    required String name,
    String color = '#2196F3',
  }) async {
    try {
      models.Tag tag;

      // If online, create on backend first to get the server-generated ID
      if (_connectivityService.isOnline) {
        try {
          final backendTag = await _apiService.createTag(
            userId: userId,
            name: name,
            color: color,
          );

          // Save backend tag to local cache
          await _database
              .into(_database.tags)
              .insertOnConflictUpdate(
                TagsCompanion.insert(
                  id: backendTag.id,
                  userId: backendTag.userId,
                  name: backendTag.name,
                  color: Value(backendTag.color),
                  createdAt: backendTag.createdAt,
                  updatedAt: backendTag.updatedAt,
                  isSynced: const Value(true),
                ),
              );

          tag = backendTag;
        } catch (e) {
          debugPrint('Failed to create tag on backend: $e');
          // Fall back to offline creation
          tag = await _createTagOffline(userId, name, color);
        }
      } else {
        // Offline - create locally
        tag = await _createTagOffline(userId, name, color);
      }

      notifyListeners();
      return tag;
    } catch (e) {
      debugPrint('Error creating tag: $e');
      rethrow;
    }
  }

  /// Create tag offline (local only)
  Future<models.Tag> _createTagOffline(
    String userId,
    String name,
    String color,
  ) async {
    final tagId = _uuid.v4();
    final now = DateTime.now();

    final tag = models.Tag(
      id: tagId,
      userId: userId,
      name: name,
      color: color,
      createdAt: now,
      updatedAt: now,
      documentCount: 0,
    );

    // Save to local cache
    await _database
        .into(_database.tags)
        .insert(
          TagsCompanion.insert(
            id: tag.id,
            userId: tag.userId,
            name: tag.name,
            color: Value(tag.color),
            createdAt: tag.createdAt,
            updatedAt: tag.updatedAt,
            isSynced: const Value(false),
          ),
        );

    return tag;
  }

  /// Update a tag
  Future<models.Tag> updateTag({
    required String tagId,
    required String userId,
    String? name,
    String? color,
  }) async {
    try {
      final now = DateTime.now();

      // If online, update on backend first
      if (_connectivityService.isOnline) {
        try {
          final updatedTag = await _apiService.updateTag(
            tagId: tagId,
            userId: userId,
            name: name,
            color: color,
          );

          // Update local cache with backend response
          await (_database.update(
            _database.tags,
          )..where((t) => t.id.equals(tagId))).write(
            TagsCompanion(
              name: Value(updatedTag.name),
              color: Value(updatedTag.color),
              updatedAt: Value(updatedTag.updatedAt),
              isSynced: const Value(true),
            ),
          );

          final documentCount = await _getDocumentCount(tagId);
          notifyListeners();

          return models.Tag(
            id: updatedTag.id,
            userId: updatedTag.userId,
            name: updatedTag.name,
            color: updatedTag.color,
            createdAt: updatedTag.createdAt,
            updatedAt: updatedTag.updatedAt,
            documentCount: documentCount,
          );
        } catch (e) {
          debugPrint('Failed to update tag on backend: $e');
          // Fall through to offline update
        }
      }

      // Offline update or backend failed
      await (_database.update(
        _database.tags,
      )..where((t) => t.id.equals(tagId))).write(
        TagsCompanion(
          name: name != null ? Value(name) : const Value.absent(),
          color: color != null ? Value(color) : const Value.absent(),
          updatedAt: Value(now),
          isSynced: const Value(false),
        ),
      );

      // Get updated tag
      final record = await (_database.select(
        _database.tags,
      )..where((t) => t.id.equals(tagId))).getSingle();

      final documentCount = await _getDocumentCount(tagId);

      notifyListeners();
      return models.Tag(
        id: record.id,
        userId: record.userId,
        name: record.name,
        color: record.color,
        createdAt: record.createdAt,
        updatedAt: record.updatedAt,
        documentCount: documentCount,
      );
    } catch (e) {
      debugPrint('Error updating tag: $e');
      rethrow;
    }
  }

  /// Delete a tag
  Future<void> deleteTag({
    required String tagId,
    required String userId,
  }) async {
    try {
      // Delete file associations
      await (_database.delete(
        _database.fileTags,
      )..where((ft) => ft.tagId.equals(tagId))).go();

      // Delete tag
      await (_database.delete(
        _database.tags,
      )..where((t) => t.id.equals(tagId))).go();

      // Sync to backend if online
      if (_connectivityService.isOnline) {
        try {
          await _apiService.deleteTag(tagId: tagId, userId: userId);
        } catch (e) {
          debugPrint('Failed to sync tag deletion to backend: $e');
          // Continue - tag is deleted locally
        }
      }

      notifyListeners();
    } catch (e) {
      debugPrint('Error deleting tag: $e');
      rethrow;
    }
  }

  /// Add a tag to a file
  Future<void> addTagToFile({
    required String userId,
    required String fileId,
    required String tagId,
  }) async {
    try {
      // Check if association already exists locally
      final existing =
          await (_database.select(_database.fileTags)..where(
                (ft) => ft.fileId.equals(fileId) & ft.tagId.equals(tagId),
              ))
              .getSingleOrNull();

      if (existing != null) {
        debugPrint('Tag already associated with file');
        return;
      }

      final fileTagId = _uuid.v4();
      final now = DateTime.now();

      // Save to local cache first (optimistic update)
      await _database
          .into(_database.fileTags)
          .insert(
            FileTagsCompanion.insert(
              id: fileTagId,
              userId: userId,
              fileId: fileId,
              tagId: tagId,
              createdAt: now,
              isSynced: const Value(false),
            ),
          );

      // Sync to backend if online
      if (_connectivityService.isOnline) {
        try {
          await _apiService.addTagToFile(
            userId: userId,
            fileId: fileId,
            tagId: tagId,
          );
          await _markFileTagSynced(fileTagId);
        } catch (e) {
          debugPrint('Failed to sync file tag to backend: $e');
          // Continue - will sync later
        }
      }

      notifyListeners();
    } catch (e) {
      debugPrint('Error adding tag to file: $e');
      rethrow;
    }
  }

  /// Remove a tag from a file
  Future<void> removeTagFromFile({
    required String userId,
    required String fileId,
    required String tagId,
  }) async {
    try {
      // Delete from local cache
      await (_database.delete(
        _database.fileTags,
      )..where((ft) => ft.fileId.equals(fileId) & ft.tagId.equals(tagId))).go();

      // Sync to backend if online
      if (_connectivityService.isOnline) {
        try {
          await _apiService.removeTagFromFile(
            userId: userId,
            fileId: fileId,
            tagId: tagId,
          );
        } catch (e) {
          debugPrint('Failed to sync file tag removal to backend: $e');
          // Continue - removed locally
        }
      }

      notifyListeners();
    } catch (e) {
      debugPrint('Error removing tag from file: $e');
      rethrow;
    }
  }

  /// Get tags for a file
  Future<List<models.Tag>> getTagsForFile(String fileId) async {
    try {
      // Get file-tag associations
      final fileTagRecords = await (_database.select(
        _database.fileTags,
      )..where((ft) => ft.fileId.equals(fileId))).get();

      if (fileTagRecords.isEmpty) {
        return [];
      }

      final tagIds = fileTagRecords.map((ft) => ft.tagId).toList();

      // Get tag details
      final tagRecords = await (_database.select(
        _database.tags,
      )..where((t) => t.id.isIn(tagIds))).get();

      return tagRecords
          .map(
            (record) => models.Tag(
              id: record.id,
              userId: record.userId,
              name: record.name,
              color: record.color,
              createdAt: record.createdAt,
              updatedAt: record.updatedAt,
            ),
          )
          .toList();
    } catch (e) {
      debugPrint('Error getting tags for file: $e');
      rethrow;
    }
  }

  /// Bulk tag files
  Future<void> bulkTagFiles({
    required String userId,
    required List<String> fileIds,
    required List<String> tagIds,
  }) async {
    try {
      for (final fileId in fileIds) {
        for (final tagId in tagIds) {
          await addTagToFile(userId: userId, fileId: fileId, tagId: tagId);
        }
      }

      notifyListeners();
    } catch (e) {
      debugPrint('Error bulk tagging files: $e');
      rethrow;
    }
  }

  /// Sync tags from backend
  Future<void> _syncTagsFromBackend() async {
    try {
      // Get current user from auth service
      final user = _authService.currentUser;
      if (user == null) {
        debugPrint('No user authenticated, skipping backend sync');
        return;
      }

      final tags = await _apiService.getTags(userId: user.id);

      // Get local tag IDs to detect deletions
      final localTags = await _database.select(_database.tags).get();
      final localTagIds = localTags.map((t) => t.id).toSet();
      final backendTagIds = tags.map((t) => t.id).toSet();

      // Upsert backend tags to local cache
      for (final tag in tags) {
        await _database
            .into(_database.tags)
            .insertOnConflictUpdate(
              TagsCompanion.insert(
                id: tag.id,
                userId: tag.userId,
                name: tag.name,
                color: Value(tag.color),
                createdAt: tag.createdAt,
                updatedAt: tag.updatedAt,
                isSynced: const Value(true),
              ),
            );
      }

      // Remove local tags that don't exist on backend (deleted elsewhere)
      final deletedTagIds = localTagIds.difference(backendTagIds);
      for (final tagId in deletedTagIds) {
        await (_database.delete(
          _database.tags,
        )..where((t) => t.id.equals(tagId))).go();
      }
    } catch (e) {
      debugPrint('Error syncing tags from backend: $e');
      // Don't rethrow - continue with local cache
    }
  }

  /// Mark tag as synced
  Future<void> _markTagSynced(String tagId) async {
    await (_database.update(_database.tags)..where((t) => t.id.equals(tagId)))
        .write(const TagsCompanion(isSynced: Value(true)));
  }

  /// Mark file tag as synced
  Future<void> _markFileTagSynced(String fileTagId) async {
    await (_database.update(_database.fileTags)
          ..where((ft) => ft.id.equals(fileTagId)))
        .write(const FileTagsCompanion(isSynced: Value(true)));
  }

  /// Get document count for a tag
  Future<int> _getDocumentCount(String tagId) async {
    final count =
        await (_database.selectOnly(_database.fileTags)
              ..addColumns([_database.fileTags.id.count()])
              ..where(_database.fileTags.tagId.equals(tagId)))
            .getSingle();

    return count.read(_database.fileTags.id.count()) ?? 0;
  }

  /// Sync unsynced tags to backend (call when coming back online)
  Future<void> syncUnsyncedTags() async {
    if (!_connectivityService.isOnline) {
      return;
    }

    try {
      final user = _authService.currentUser;
      if (user == null) {
        return;
      }

      // Get unsynced tags
      final unsyncedTags = await (_database.select(
        _database.tags,
      )..where((t) => t.isSynced.equals(false))).get();

      for (final tag in unsyncedTags) {
        try {
          // Try to create on backend
          await _apiService.createTag(
            userId: user.id,
            name: tag.name,
            color: tag.color,
          );

          // Mark as synced
          await _markTagSynced(tag.id);
        } catch (e) {
          debugPrint('Failed to sync tag ${tag.id}: $e');
          // Continue with next tag
        }
      }

      // Get unsynced file tags
      final unsyncedFileTags = await (_database.select(
        _database.fileTags,
      )..where((ft) => ft.isSynced.equals(false))).get();

      for (final fileTag in unsyncedFileTags) {
        try {
          await _apiService.addTagToFile(
            userId: user.id,
            fileId: fileTag.fileId,
            tagId: fileTag.tagId,
          );

          await _markFileTagSynced(fileTag.id);
        } catch (e) {
          debugPrint('Failed to sync file tag ${fileTag.id}: $e');
          // Continue with next file tag
        }
      }

      notifyListeners();
    } catch (e) {
      debugPrint('Error syncing unsynced tags: $e');
    }
  }
}
